cmd/tx.go

```diff
@@ -40,31 +42,105 @@ var (
 			rawTx, err := ioutil.ReadFile(filename)
 			cobra.CheckErr(err)
 
-			// Determine what kind of a transaction this is by attempting to decode it as either a
-			// consensus layer transaction or a runtime transaction.
-			var (
-				sigTx  interface{}
-				consTx consensusTx.SignedTransaction
-				rtTx   types.UnverifiedTransaction
-			)
-			if err = json.Unmarshal(rawTx, &consTx); err != nil {
-				// Failed to unmarshal as consensus transaction, try as runtime transaction.
-				if err = json.Unmarshal(rawTx, &rtTx); err != nil {
-					cobra.CheckErr(fmt.Errorf("malformed transaction"))
-				}
-				sigTx = &rtTx
-			} else {
-				sigTx = &consTx
+			tx, err := tryDecodeTx(rawTx)
+			cobra.CheckErr(err)
+
+			var sigTx, meta interface{}
+			switch dtx := tx.(type) {
+			case *consensusTx.SignedTransaction, *types.UnverifiedTransaction:
+				// Signed transaction, just broadcast.
+				sigTx = tx
+			case *consensusTx.Transaction:
+				// Unsigned consensus transaction, sign first.
+				acc := common.LoadAccount(cfg, npa.AccountName)
+				sigTx, err = common.SignConsensusTransaction(ctx, npa, acc, conn, dtx)
+				cobra.CheckErr(err)
+			case *types.Transaction:
+				// Unsigned runtime transaction, sign first.
+				acc := common.LoadAccount(cfg, npa.AccountName)
+				sigTx, meta, err = common.SignParaTimeTransaction(ctx, npa, acc, conn, dtx)
+				cobra.CheckErr(err)
 			}
 
 			// Broadcast signed transaction.
-			common.BroadcastTransaction(ctx, npa.ParaTime, conn, sigTx, nil, nil)
+			common.BroadcastTransaction(ctx, npa.ParaTime, conn, sigTx, meta, nil)
+		},
+	}
+
+	txShowCmd = &cobra.Command{
+		Use:   "show <filename.json>",
+		Short: "Pretty print a transaction",
+		Args:  cobra.ExactArgs(1),
+		Run: func(cmd *cobra.Command, args []string) {
+			cfg := cliConfig.Global()
+			npa := common.GetNPASelection(cfg)
+			filename := args[0]
+
+			rawTx, err := ioutil.ReadFile(filename)
+			cobra.CheckErr(err)
+
+			tx, err := tryDecodeTx(rawTx)
+			cobra.CheckErr(err)
+
+			common.PrintTransaction(npa, tx)
 		},
 	}
 )
 
+func tryDecodeTx(rawTx []byte) (any, error) {
```

Comment:
Should this go into oasis-sdk? E.g. to simplify the runtime test vectors and for debugging future client-sdk/go apps?

Reply:
I would rather not bloat the SDK until this is required somewhere else as well.